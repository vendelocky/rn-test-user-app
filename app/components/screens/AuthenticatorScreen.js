import React, { useState, useRef } from "react";
import {
  View,
  Text,
  StyleSheet,
  TextInput,
  TouchableOpacity,
  Alert,
  ActivityIndicator,
  BackHandler,
} from "react-native";
import Icon from "react-native-vector-icons/FontAwesome";
import Spacer from "../Spacer";
import text from "../../constants/text";
import type from "../../constants/type";
import { useNavigation } from "@react-navigation/core";
import fetchData from "../fetchData";
import { useFocusEffect } from "@react-navigation/native";
import exitAlert from "../exitAlert";
import StringifyAsyncStorage from "../StringifyAsyncStorage";
import storageKey from "../../constants/storageKey";
import pageName from "../../constants/pageName";
import colors from "../../constants/colors";

const AuthenticatorScreen = () => {
  const [changeAuthPage, setChangeAuthPage] = useState(type.LOGIN);
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [loading, setLoading] = useState(false);
  const emailTextInput = useRef();
  const passwordTextInput = useRef();
  const inputErrorMessage = useRef("");
  const navigation = useNavigation();

  useFocusEffect(
    React.useCallback(() => {
      const onBackPress = () => {
        exitAlert();
        return true;
      };

      BackHandler.addEventListener("hardwareBackPress", onBackPress);

      return () =>
        BackHandler.removeEventListener("hardwareBackPress", onBackPress);
    }, [])
  );

  const onPressButton = async (page) => {
    if (validationChecking()) {
      const details = {
        email: email.toLowerCase(),
        password: password,
      };
      setLoading(true);
      await fetchData(page, details)
        .then((response) => response.json())
        .then((responseJson) => {
          setLoading(false);
          if (responseJson.error) {
            Alert.alert(responseJson.error);
          } else {
            clearInput();
            if (page === type.LOGIN) {
              StringifyAsyncStorage.setItem(
                storageKey.TOKEN,
                responseJson?.token
              );
              navigation.navigate(pageName.userList);
            } else {
              Alert.alert(text.registerSuccessTitle, text.registerSuccessBody);
              setChangeAuthPage(type.LOGIN);
            }
          }
        })
        .catch((error) => {
          setLoading(false);
          Alert.alert(error);
          console.error(error);
        });
    } else {
      Alert.alert(inputErrorMessage.current);
    }
  };

  const changePage = () => {
    clearInput();
    setChangeAuthPage((prevState) => {
      if (prevState === type.LOGIN) {
        return type.REGISTER;
      }
      return type.LOGIN;
    });
  };

  const clearInput = () => {
    setEmail("");
    setPassword("");
    emailTextInput.current.clear();
    passwordTextInput.current.clear();
  };

  const validationChecking = () => {
    if (!email || !password) {
      inputErrorMessage.current = !email ? text.emptyEmail : text.emptyPassword;
      return false;
    }

    const emailRegex = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/;
    if (!emailRegex.test(email)) {
      inputErrorMessage.current = text.invalidEmail;
      return false;
    }
    return true;
  };

  return (
    <View style={styles.root}>
      <Text style={styles.title}>{text.title}</Text>
      <Spacer height={44} />
      <View style={styles.inputContainer}>
        <View style={styles.inputIcon}>
          <Icon name="envelope" size={20} color={colors.black} />
        </View>
        <TextInput
          ref={emailTextInput}
          style={styles.input}
          placeholder={text.email}
          placeholderTextColor={colors.grey}
          keyboardType="email-address"
          onChangeText={(emailText) => setEmail(emailText)}
        />
      </View>
      <Spacer height={24} />
      <View style={styles.inputContainer}>
        <View style={styles.inputIcon}>
          <Icon name="lock" size={20} color={colors.black} />
        </View>
        <TextInput
          ref={passwordTextInput}
          style={styles.input}
          placeholder={text.password}
          placeholderTextColor={colors.grey}
          secureTextEntry={true}
          onChangeText={(passwordText) => setPassword(passwordText)}
        />
      </View>
      <Spacer height={40} />
      <TouchableOpacity
        onPress={() => onPressButton(changeAuthPage)}
        style={styles.button}
      >
        <Text style={styles.buttonText}>
          {changeAuthPage === type.LOGIN ? text.login : text.register}
        </Text>
      </TouchableOpacity>
      <Spacer height={56} />
      <View>
        <Text style={styles.footerText}>
          {changeAuthPage === type.LOGIN
            ? text.loginAccount
            : text.registerAccount}
          <Spacer width={12} />
          <Text style={styles.link} onPress={changePage}>
            {changeAuthPage === type.LOGIN ? text.register : text.login}
          </Text>
        </Text>
      </View>
      {loading && (
        <ActivityIndicator
          style={styles.loading}
          size={"large"}
          color={colors.blue}
        />
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  root: {
    flex: 1,
    alignItems: "center",
    alignSelf: "stretch",
    backgroundColor: colors.white,
  },

  title: {
    fontSize: 24,
    marginTop: 56,
    color: colors.black,
  },

  inputContainer: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: colors.white,
    borderWidth: 1,
    borderRadius: 5,
    borderColor: colors.grey,
    marginHorizontal: 48,
  },

  inputIcon: {
    padding: 12,
    width: 48,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: colors.grey,
  },

  input: {
    flex: 1,
    height: 44,
    paddingLeft: 12,
    color: colors.black,
  },

  button: {
    backgroundColor: colors.blue,
    width: 156,
    height: 44,
    borderRadius: 5,
    justifyContent: "center",
    alignItems: "center",
  },

  buttonText: {
    color: colors.white,
    fontSize: 16,
  },

  footerText: {
    fontSize: 16,
    color: colors.black,
  },

  link: {
    color: colors.blue,
    textDecorationLine: "underline",
  },

  loading: {
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    alignItems: "center",
    justifyContent: "center",
    zIndex: 2,
    backgroundColor: colors.white,
    opacity: 0.7,
  },
});

export default AuthenticatorScreen;
