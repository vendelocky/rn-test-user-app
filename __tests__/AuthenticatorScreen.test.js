/**
 * @format
 */

import React from "react";
import { cleanup, render } from "@testing-library/react-native";
import AuthenticatorScreen from "../app/components/screens/AuthenticatorScreen";
import MockedNavigator from "../app/navigations/MockedNavigator";
import renderer from "react-test-renderer";
import { TextInput, TouchableOpacity } from "react-native";
import Icon from "react-native-vector-icons/FontAwesome";

const mockedNavigation = jest.fn();
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useFocusEffect: () => mockedNavigation,
    useNavigation: () => ({
      navigate: mockedNavigation,
    }),
  };
});

describe("Authenticator Component", () => {
  afterEach(cleanup);

  it("should match the snapshot", () => {
    const { toJSON } = render(<MockedNavigator component={AuthenticatorScreen} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it("renders correctly", () => {
    renderer.create(<MockedNavigator component={AuthenticatorScreen} />);
  });

  const testRenderer = renderer.create(<MockedNavigator component={AuthenticatorScreen} />);
  const testInstance = testRenderer.root;

  describe("Text Input Component", () => {
    const textInput = testInstance.findAllByType(TextInput);
    it("should be defined", () => {
      expect(textInput).toBeDefined();
    });

    it("has 2 text inputs", () => {
      expect(textInput).toHaveLength(2);
    });
  });

  describe("Icon Component", () => {
    const icon = testInstance.findAllByType(Icon);
    it("has 2 icons", () => {
      expect(icon).toHaveLength(2);
    });
  });

  describe("Button Component", () => {
    const button = testInstance.findByType(TouchableOpacity);
    it("should be defined", () => {
      expect(button).toBeDefined();
    });

    it("has text", () => {
      const buttonText = button.children;
      expect(buttonText).not.toBeNull();
    });
  });
});
