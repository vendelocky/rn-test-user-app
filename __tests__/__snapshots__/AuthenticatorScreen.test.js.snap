// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authenticator Component should match the snapshot 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <RNSScreenStack
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNSScreen
      collapsable={false}
      forwardedRef={[Function]}
      nativeID="animatedComponent"
      onAppear={[Function]}
      onDisappear={[Function]}
      onDismissed={[Function]}
      onWillDisappear={[Function]}
      replaceAnimation="push"
      stackPresentation="push"
      style={
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <RNSScreenStackHeaderConfig
        backButtonInCustomView={false}
        backgroundColor="rgb(255, 255, 255)"
        color="rgb(0, 122, 255)"
        direction="ltr"
        disableBackButtonMenu={false}
        hidden={false}
        hideBackButton={false}
        hideShadow={false}
        largeTitleHideShadow={false}
        title="MockedScreen"
        titleColor="rgb(28, 28, 30)"
        topInsetEnabled={false}
        translucent={false}
      />
      <View
        style={
          Array [
            Object {
              "flex": 1,
            },
            Object {
              "backgroundColor": "rgb(242, 242, 242)",
            },
            undefined,
          ]
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "alignSelf": "stretch",
              "backgroundColor": "#ffffff",
              "flex": 1,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#000000",
                "fontSize": 24,
                "marginTop": 56,
              }
            }
          >
            UsersApp
          </Text>
          <View
            style={
              Array [
                null,
                Object {
                  "height": 44,
                },
              ]
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderColor": "#cccccc",
                "borderRadius": 5,
                "borderWidth": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginHorizontal": 48,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#cccccc",
                  "justifyContent": "center",
                  "padding": 12,
                  "width": 48,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <TextInput
              keyboardType="email-address"
              onChangeText={[Function]}
              placeholder="Email"
              placeholderTextColor="#cccccc"
              style={
                Object {
                  "color": "#000000",
                  "flex": 1,
                  "height": 44,
                  "paddingLeft": 12,
                }
              }
            />
          </View>
          <View
            style={
              Array [
                null,
                Object {
                  "height": 24,
                },
              ]
            }
          />
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderColor": "#cccccc",
                "borderRadius": 5,
                "borderWidth": 1,
                "flexDirection": "row",
                "justifyContent": "center",
                "marginHorizontal": 48,
              }
            }
          >
            <View
              style={
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#cccccc",
                  "justifyContent": "center",
                  "padding": 12,
                  "width": 48,
                }
              }
            >
              <Text
                allowFontScaling={false}
                style={
                  Array [
                    Object {
                      "color": "#000000",
                      "fontSize": 20,
                    },
                    undefined,
                    Object {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    Object {},
                  ]
                }
              >
                
              </Text>
            </View>
            <TextInput
              onChangeText={[Function]}
              placeholder="Password"
              placeholderTextColor="#cccccc"
              secureTextEntry={true}
              style={
                Object {
                  "color": "#000000",
                  "flex": 1,
                  "height": 44,
                  "paddingLeft": 12,
                }
              }
            />
          </View>
          <View
            style={
              Array [
                null,
                Object {
                  "height": 40,
                },
              ]
            }
          />
          <View
            accessible={true}
            collapsable={false}
            focusable={true}
            nativeID="animatedComponent"
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#4099da",
                "borderRadius": 5,
                "height": 44,
                "justifyContent": "center",
                "opacity": 1,
                "width": 156,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "fontSize": 16,
                }
              }
            >
              Login
            </Text>
          </View>
          <View
            style={
              Array [
                null,
                Object {
                  "height": 56,
                },
              ]
            }
          />
          <View>
            <Text
              style={
                Object {
                  "color": "#000000",
                  "fontSize": 16,
                }
              }
            >
              Don't have account yet?
              <View
                style={
                  Array [
                    Object {
                      "width": 12,
                    },
                    null,
                  ]
                }
              />
              <Text
                onPress={[Function]}
                style={
                  Object {
                    "color": "#4099da",
                    "textDecorationLine": "underline",
                  }
                }
              >
                Register
              </Text>
            </Text>
          </View>
        </View>
      </View>
    </RNSScreen>
  </RNSScreenStack>
</RNCSafeAreaProvider>
`;
